design/design-doc-flexible-claims-and-composition.md

```diff
@@ -0,0 +1,945 @@
+            matchComposite: true
+          autoscaling:
+            enabled: true
+            minNodeCount: 0
+            maxNodecount: 100
+          config:
+            diskType: PD_SSD
+            machineType: n1-highmem-32
+          version: 1.16
+        providerRef:
+          name: example
+        reclaimPolicy: Delete
+    patch:
+    - fromClaimFieldPath: "spec.version"
+      toResourceFieldPath: "spec.forProvider.version"
+```
+
+### Transform Functions
+
+TODO(negz): Explain transform functions.
+
+* Simple, schemafied, functions that can transform claim fields.
+* Similar to Terraform's small set of functions.
+* Purposefully not open ended.
+* Written using plain old YAML.
+
+### Composite References
+
+Crossplane allows certain fields of primitive managed resources to be set to a
+value inferred from another primitive managed resource using [cross resource
+references]. Take for example a resource that must specify the VPC network in
+which it should be created, by specifying the `network` field:
+
+```yaml
+spec:
+  forProvider:
+    network: /projects/example/global/networks/desired-vpc-network
+```
+
+Cross resource references allow this resource to instead reference the `Network`
+managed resource that represents the desired VPC Network:
+
+```yaml
+spec:
+  forProvider:
+    networkRef:
+      name: desired-vpc-network
+```
+
+The managed resource reconcile logic resolves this reference and populates the
+`network` field, resulting in the following configuration:
+
+```yaml
+spec:
+  forProvider:
+    # Network is populated with the value calculated by networkRef.
+    network: /projects/example/global/networks/desired-vpc-network
+    networkRef:
+      name: desired-vpc-network
+```
+
+This functionality must be extended in order to support composite resource
+classes. Consider a `CompositeResourceClass` that may be used to dynamically
+provision the following primitive managed resources:
+
+* `Subnetwork` A
+* `GKECluster` A
+* `ServiceAccount` A
+* `ServiceAccount` B
+* `GKENodePool` A
+* `GKENodePool` B
+
+The `CompositeResourceClass` author would like to configure the resources such
+that:
+
+1. `Subnetwork` A is created in an existing, statically provisioned `Network`.
+1. `GKECluster` A is created in `Subnetwork` A.
+1. `GKENodePool` A uses `ServiceAccount` A.
+1. `GKENodePool` B uses `ServiceAccount` B.
+1. Both `GKENodePool` resources join `GKECluster` A.
+
+The author cannot use a contemporary cross resource reference for requirements
+two through five. Managed resources are referenced by name, and the names of
+dynamically provisioned resources are non-deterministic; they are not known
+until they have been provisioned.
+
+This document proposes the introduction of a reference _selector_, which allows
+a managed resource to describe the properties of the distinct resource it wishes
+to reference, rather than explicitly naming it.
+
+```yaml
+spec:
+  forProvider:
+    networkSelector:
```

Comment:
Will `networkRef` still exist?

Reply:
Yeah - the idea is `networkSelector` would set `networkRef`. Similar to how a claim's `classSelector` sets its `classRef`.